cpp/include/raft/core/detail/copy.hpp

```diff
@@ -0,0 +1,524 @@
+  auto static check_for_unique_dst(dst_type dst)
+  {
+    if constexpr (!dst_type::is_always_unique()) {
+      RAFT_EXPECTS(dst.is_unique(), "Destination mdspan must be unique for parallelized copies");
+    }
+  }
+
+  auto static constexpr const src_contiguous =
+    std::disjunction_v<std::is_same<src_layout_type, layout_c_contiguous>,
+                       std::is_same<src_layout_type, layout_f_contiguous>>;
+
+  auto static constexpr const dst_contiguous =
+    std::disjunction_v<std::is_same<dst_layout_type, layout_c_contiguous>,
+                       std::is_same<dst_layout_type, layout_f_contiguous>>;
+
+  auto static constexpr const both_contiguous = src_contiguous && dst_contiguous;
+
+  auto static constexpr const same_underlying_layout =
+    std::disjunction_v<std::bool_constant<same_layout>,
+                       std::bool_constant<has_vector_rank && both_contiguous>>;
+  // Layout for intermediate tile if copying through custom kernel
+  using tile_layout_type =
+    std::conditional_t<src_contiguous,
+                       src_layout_type,
+                       std::conditional_t<dst_contiguous, dst_layout_type, layout_c_contiguous>>;
+
+  // Accessibility
+  auto static constexpr const dst_device_accessible = is_device_mdspan_v<dst_type>;
+  auto static constexpr const src_device_accessible = is_device_mdspan_v<src_type>;
+  auto static constexpr const both_device_accessible =
+    dst_device_accessible && src_device_accessible;
+
+  auto static constexpr const dst_host_accessible  = is_host_mdspan_v<dst_type>;
+  auto static constexpr const src_host_accessible  = is_host_mdspan_v<src_type>;
+  auto static constexpr const both_host_accessible = dst_host_accessible && src_host_accessible;
+
+  // Allowed copy codepaths
+  auto static constexpr const can_use_host = both_host_accessible;
+
+#if (defined(__AVX__) || defined(__SSE__) || defined(__ARM_NEON))
+  // TODO(wphicks): Following should be only necessary restrictions. Test if
+  // perf actually improves once fully implemented.
+  // auto static constexpr const can_use_simd = can_use_host && both_contiguous &&
+  // both_float_or_both_double;
+  auto static constexpr const can_use_simd =
+    can_use_host && both_contiguous && both_float && has_matrix_rank;
+#else
+  auto static constexpr const can_use_simd = false;
+#endif
+
+  auto static constexpr const can_use_std_copy =
+    std::conjunction_v<std::bool_constant<can_use_host>,
+                       std::bool_constant<compatible_dtype>,
+                       std::bool_constant<both_contiguous>,
+                       std::bool_constant<same_underlying_layout>>;
+  auto static constexpr const can_use_raft_copy =
+    std::conjunction_v<std::bool_constant<CUDA_ENABLED>,
+                       std::bool_constant<same_dtype>,
+                       std::bool_constant<both_contiguous>,
+                       std::bool_constant<same_underlying_layout>>;
+
+  auto static constexpr const requires_intermediate =
```

Comment:
Why is an intermediate needed? I am assuming this is the managed memory case

Reply:
This is for the case where we have e.g. dst on device and src on hst, _and also_ we cannot just perform a trivial cudaMemcpyAsync. In that case, we need an intermediate device buffer to store the results of the transpose or dtype conversion before we copy back to host or a to store a straight copy from the host before we perform the transpose or dtype conversion.

Reply:
Added comment explaining this to code.